docs/design/plugin_api.md

```diff
@@ -0,0 +1,363 @@
+# Plugin API
+
+## Purpose
+
+This document describes the API used by variant provider plugins.
+The plugins are a central point of the variant specification, defining
+the valid metadata, and providing routines necessary to install
+and build variants.
+
+The plugin API is defined using Python [Protocols](
+https://typing.python.org/en/latest/spec/protocol.html), as well
+as textual description.
+
+
+## Basic design
+
+Every provider plugin operates within a single namespace. The namespace
+is used as key for all plugin-related operations. All the properties
+defined by the plugin are in the plugin's namespace, and the plugin
+defines all the valid feature names and values within that namespace.
+
+Within a single package, there can be only used plugin used for a given
+namespace. Attempting to load a second plugin sharing the namespace
+must cause a fatal error. However, in the wider scope there can be
+multiple mutually exclusive plugins using the same namespace.
+For example, this could happen if a plugin becomes unmaintained
+and needs to be forked into a new package.
+
+Plugins are expected to be automatically installed into an isolated
+environment when installing packages, in order to facilitate variant
+selection. Therefore, it is necessary that the packages providing
+plugins are installable with the same indices enabled that are used
```

Comment:
I would leave index configuration to the package manager. There could e.g. be a company policy that packages can be installed from PyPI, but variant providers need to be installed from an internal index.

Reply:
Good point. Perhaps I should just leave the next sentence, i.e. that plugins for packages published on PyPI must be on PyPI as well?

Reply:
The same situation already exists with packages and their dependencies as well as packages and their build backend, so I'd just elide this: The current practices apply as you'd expect. That is not to say that the current practices around indexes don't have a lot of confusing to bad behaviors too, but it doesn't change the situation from what we already have with build backend.